src/py/flwr/cli/build.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Flower command line interface `build` command."""
+
+import hashlib
+import os
+import zipfile
+from pathlib import Path
+from typing import Optional
+
+import jwt
+import pathspec
+import typer
+from typing_extensions import Annotated
+
+from .config_utils import validate_project_dir
+from .utils import is_valid_project_name
+
+
+# pylint: disable=too-many-locals
+def build(
+    directory: Annotated[
+        Optional[Path],
+        typer.Option(help="The Flower project directory to bundle into a FAB"),
+    ] = None,
+    signed: Annotated[bool, typer.Option(help="Flag to sign the FAB")] = False,
+) -> None:
+    """Build a Flower project into a FAB file.
+
+    It can be ran without any argument:
+
+        `flwr build`
+
+    In which case, it will try to bundle the current directory if it is a
+    valid Flower App.
+
+    It can also be ran with a directory argument:
+
+        `flwr build --directory ./docs/target_dir`
+
+    And finally, the `--signed` argument can also be used to sign the FAB:
+
+        `flwr build --signed`
+
+    Note that this will prompt the user for its private key.
+    """
+    if directory is None:
+        directory = Path.cwd()
+
+    directory = directory.resolve()
+    if not directory.is_dir():
+        typer.secho(
+            f"❌ The path {directory} is not a valid directory.",
+            fg=typer.colors.RED,
+            bold=True,
+        )
+        raise typer.Exit(code=1)
+
+    if not is_valid_project_name(directory.name):
+        typer.secho(
+            f"❌ The project name {directory.name} is invalid, "
+            "a valid project name must start with a letter or an underscore, "
+            "and can only contain letters, digits, and underscores.",
+            fg=typer.colors.RED,
+            bold=True,
+        )
+        raise typer.Exit(code=1)
```

Comment:
This is too restrictive. The directory name of the project should allow `-` (quite common in Python library names). The Python package inside will then use `_` instead of `-` to make it a valid Python package name.
Example:
- Project name: `flwr-datasets` (e.g., `pip install flwr-datasets`)
- Package name: `flwr_datasets`

Reply:
We can already make the change here. All projects generated by `flwr new` would be compatible.

Reply:
Here is the PR: #3309